filebeat/tests/integration/filestream_test.go

```diff
@@ -657,6 +671,471 @@ func TestFilestreamTakeOverFromLogInput(t *testing.T) {
+			filebeat.WriteConfigFile(cfgYAML)
+			filebeat.Start()
+
+			for _, msgFmt := range tc.msgs {
+				msg := fmt.Sprintf(msgFmt, msgLogFilePath)
+				filebeat.WaitForLogs(
+					msg,
+					10*time.Second,
+					"did not find '%s' in the logs",
+					msg,
+				)
+			}
+
+			if tc.resourceNotFinished {
+				// Wait a few times for the 'not finished' logs
+				notFinishedMsg := fmt.Sprintf(
+					"not all events from '%s' have been published, "+
+						"closing harvester",
+					msgLogFilePath)
+				for i := range 2 {
+					filebeat.WaitForLogs(
+						notFinishedMsg,
+						10*time.Second,
+						"[%d] Filebeat did not wait for the resource to be finished",
+						i,
+					)
+				}
+
+				if tc.dataAdded {
+					// Add more data to the log file
+					integration.GenerateLogFile(t, msgLogFilePath, 5, true)
+
+					// Wait for the "file update" log
+					filebeat.WaitForLogs(
+						fmt.Sprintf("File %s has been updated", msgLogFilePath),
+						time.Second,
+						"filewatcher did not detect the file as updated")
+
+					// Wait for the harvester to be closed
+					filebeat.WaitForLogs(
+						fmt.Sprintf("not all events from '%s' have been published, closing harvester", msgLogFilePath),
+						10*time.Second,
+						"harvester was not closed after data added to the file")
+
+					// Wait for the "not changed" log
+					filebeat.WaitForLogs(
+						fmt.Sprintf("File %s has not changed, trying to start new harvester because 'delete' is enabled", msgLogFilePath),
+						time.Second,
+						"Filestream did not try to start a new harvester for the unchanged file")
+
+					// Ensure harvester closes without removing the file
+					filebeat.WaitForLogs(
+						fmt.Sprintf("not all events from '%s' have been published, closing harvester", msgLogFilePath),
+						10*time.Second,
+						"Harvester was not closed because the resource is not finished")
+
+					if !fileExists(t, logFile) {
+						t.Fatalf("%q should not have been removed", logFile)
+					}
+				}
+
+				if err := es.UpdateOdds(0, 0, 0, 0); err != nil {
+					t.Fatalf("cannot update mock-es odds: %s", err)
+				}
+			}
+
+			msg := fmt.Sprintf("'%s' removed", msgLogFilePath)
+			filebeat.WaitForLogs(msg, 30*time.Second, "file removed log entry not found")
+			removedMsg := filebeat.GetLastLogLine(msg)
+
+			gracePeriodMsg := fmt.Sprintf("all events from '%s' have been published, waiting for %s grace period", msgLogFilePath, tc.gracePeriod)
+			beforeWait := filebeat.GetLastLogLine(gracePeriodMsg)
+
+			delta := timeBetweenLogEntries(t, beforeWait, removedMsg)
+			if delta < tc.gracePeriod {
+				t.Errorf("grace period of %s was not respected", tc.gracePeriod)
+				t.Log("grace period waiting calculated based on the following log entries:")
+				t.Log("First :", beforeWait)
+				t.Log("Second:", removedMsg)
+			}
+
+			if fileExists(t, logFile) {
+				t.Fatalf("%q should have been removed", logFile)
+			}
+		})
+	}
+}
+
+func TestFilestreamDeleteRestart(t *testing.T) {
```

Comment:
Is this test necessary? Isn't it just testing that Filebeat doesn't delete the file if Filebeat isn't running? Isn't that true by definition?

Also there is he chance for this to be a little bit race-y if the grace period manages to elapse while Filebeat hasn't completely shut down, then the file might be correctly deleted depending on the timing.

I can't tell if 5 seconds is always enough time without looking at a measurement of how fast this test completes on different CI workers.